packages/components/datepicker/src/Datepicker.tsx

```diff
@@ -1,27 +1,177 @@
+const DATE_FORMAT = 'dd/MM/yyyy';
 
-function _Datepicker(props: DatepickerProps, ref: React.Ref<HTMLDivElement>) {
+export type DatepickerProps = CommonProps & {
+  onSelect: (day: Date | undefined) => void;
+} & Omit<
+    DayPickerSingleProps,
+    'mode' | 'onSelect' | 'fromMonth' | 'toMonth' | 'fromYear' | 'toYear'
+  > &
+  Pick<
+    TextInputProps,
+    'id' | 'isDisabled' | 'isInvalid' | 'isReadOnly' | 'isRequired'
+  > &
+  Pick<PopoverProps, 'usePortal'>;
+
+/**
+ * Provides functionality for date selection.
+ */
+export function Datepicker(props: DatepickerProps) {
   const styles = getStyles();
+  const {
+    //common props
+    testId = 'cf-datepicker',
+    className,
+    style,
+    // input props
+    id,
+    isDisabled,
+    isInvalid,
+    isReadOnly,
+    isRequired,
+    // popover props
+    usePortal,
+    // calendar props
+    selected,
+    onSelect,
+    fromDate,
+    toDate,
+    locale,
+    ...restCalendarProps
+  } = props;
+
+  const isDateValid = useCallback(
+    (date: Date) => {
+      if (!isValid(date)) {
+        return false;
+      }
+      if (fromDate && startOfDay(fromDate).getTime() > date.getTime()) {
+        return false;
+      }
+      if (toDate && date.getTime() > endOfDay(toDate).getTime()) {
+        return false;
+      }
+      return true;
+    },
+    [fromDate, toDate],
+  );
+
+  const parseInputDate = useCallback(
+    (value: string) => parse(value, DATE_FORMAT, new Date(), { locale }),
+    [locale],
+  );
+
+  const [isPopoverOpen, setIsPopoverOpen] = useState(false);
+  const [inputValue, setInputValue] = useState<string>(() =>
+    selected ? format(selected, DATE_FORMAT) : '',
+  );
+
+  useEffect(() => {
+    if (
+      selected &&
+      selected.getTime() !== parseInputDate(inputValue).getTime()
+    ) {
+      setInputValue(format(selected, DATE_FORMAT));
+    }
+    // we want to run this hook only when `selected` prop changes
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, [selected]);
+
+  const handleInputChange: ChangeEventHandler<HTMLInputElement> = useCallback(
+    (e) => {
+      setInputValue(e.currentTarget.value);
+
+      const date = parseInputDate(e.currentTarget.value);
+      if (isDateValid(date)) {
+        onSelect(date);
+      } else {
+        onSelect(undefined);
+      }
+    },
+    [onSelect, parseInputDate, isDateValid],
+  );
+
+  const handleDaySelect = useCallback(
+    (date: Date) => {
+      if (date && isDateValid(date)) {
+        onSelect(date);
+        setIsPopoverOpen(false);
+      }
+    },
+    [onSelect, isDateValid],
+  );
+
+  const isTextInputValueInvalid =
+    inputValue && !isDateValid(parseInputDate(inputValue));
 
   return (
-    <div
-      data-test-id={props.testId}
-      ref={ref}
-      className={cx(styles.datepicker, props.className)}
+    <Popover
+      isOpen={isPopoverOpen}
+      onClose={() => setIsPopoverOpen(false)}
+      usePortal={usePortal}
```

Comment:
this will be undefined by default, so we might need to just to spread the props here.

Reply:
I made it like this to prevent bleeding it to the `Calendar` component through `otherProps`